subprojects/core/src/main/java/org/gradle/process/internal/JavaExecHandleBuilder.java

```diff
@@ -306,58 +313,52 @@ private ConfigurableFileCollection doGetClasspath() {
+                }
             }
         }
 
-        return new ArgsEnvironment(arguments, environment);
+        return arguments;
     }
 
-    private boolean shouldUseEnvironmentVariable(List<String> arguments, Map<String, String> environment) {
+    private boolean shouldTryShorteningCommandLine(List<String> arguments) {
         int classPathFlagIndex = arguments.indexOf("-cp");
         if (classPathFlagIndex == -1) {
             // No class path flag, so we can't use CLASSPATH
             LOGGER.info("No class path flag found in the command line, Gradle cannot shorten the command line");
             return false;
         }
 
-        if (hasEnvironmentVariableExceedMaxLength(arguments.get(classPathFlagIndex + 1))) {
-            // Class path is exceeding the maximum environment variable value length
-            LOGGER.info("Class path size is exceeding the maximum environment variable length, CLASSPATH variable cannot be used for shortening the command line");
-            return false;
-        }
-
-        if (System.getenv().containsKey("CLASSPATH")) {
-            if (environment.containsKey("CLASSPATH")) {
-                // Only if the CLASSPATH wasn't overwritten in the JavaExecSpec
-                if (System.getenv().get("CLASSPATH").equals(environment.get("CLASSPATH"))) {
-                    return true;
-                }
-                LOGGER.info("CLASSPATH environment variable was explicitly overwritten, Gradle cannot shorten the command line");
-                return false;
-            }
-            // The CLASSPATH was explicitly cleared
-            LOGGER.info("CLASSPATH environment variable was explicitly cleared, Gradle cannot shorten the command line");
-            return false;
-        }
+        return true;
+    }
 
-        if (environment.containsKey("CLASSPATH")) {
-            // The CLASSPATH was explicitly defined
-            LOGGER.info("CLASSPATH environment variable was explicitly defined, Gradle cannot shorten the command line");
-            return false;
+    private List<String> writePathingJarFile(FileCollection classPath, File pathingJarFile) throws IOException {
+        try (JarOutputStream jarOutputStream = new JarOutputStream(new FileOutputStream(pathingJarFile), toManifest(classPath))) {
+            jarOutputStream.putNextEntry(new ZipEntry("META-INF/"));
         }
+        return Arrays.asList("-cp", pathingJarFile.getAbsolutePath());
+    }
 
-        // The CLASSPATH wasn't declared or inherited
-        return true;
+    private static Manifest toManifest(FileCollection classPath) {
+        Manifest manifest = new Manifest();
+        Attributes attributes = manifest.getMainAttributes();
+        attributes.put(Attributes.Name.MANIFEST_VERSION, "1.0");
+        attributes.putValue("Class-Path", classPath.getFiles().stream().map(File::toURI).map(URI::toString).collect(Collectors.joining(" ")));
```

Comment:
We can't really refactor the current classpath as a string because we need to transform all `File` to `URI`.